### PR TITLE
Adds a catalog, reading of manifest and manifest_list files and a Tableprovider for Apache Datafusion

### PR DESCRIPTION
**Pull request by JanKaul**
# Purpose of this PR
The final goal of this PR is to enable reading iceberg tables with Apache Datafusion. For that I completed the implementation to read manifest files and added support to read manifest_list files. As a next step I created an iterator that iterates over all files in the iceberg table.
I'm currently working on the Datafusion reader.

### CODE REVIEW

**Comment by oliverdaff** [comment]
It's fantastic to see such a comprehensive PR.  I'm working through the review, please be patient while I carve out some time to give this the attention it deserves.

**Comment by JanKaul** [comment]
Hi @oliverdaff, thanks for the reply. I'm currently implementing an [Arrow Datafusion](https://arrow.apache.org/datafusion/user-guide/introduction.html) table provider for iceberg tables in another feature branch. In that I also added some improvements and fixes for the manifests and manifest_list modules. Including proper support for v1 and v2 manifests and files. I would suggest you wait reviewing the changes until I merge the datafusion branch into my master branch. I will also improve documentation and tests.

My code should just be seen as a starting point, I'm open to discussions and refactoring the code.

File: src/model/snapshot.rs [hunk 1]
```diff
@@ -64,6 +64,46 @@ pub struct SnapshotV2 {
...
+}
+
+impl From<SnapshotV1> for SnapshotV2 {
+    fn from(v1: SnapshotV1) -> Self {
+        SnapshotV2 {
+            snapshot_id: v1.snapshot_id,
+            parent_snapshot_id: v1.parent_snapshot_id,
+            sequence_number: 0,
+            timestamp_ms: v1.timestamp_ms,
+            manifest_list: v1.manifest_list.unwrap_or_else(|| "".to_owned()),
```
Comment (oliverdaff) [inline src/model/snapshot.rs:97]: Consider using `unwrap_or_default` here.

**Review by oliverdaff — COMMENTED** [review]
This is a very large PR so I'm going through it in modules as I have time.

I've reviewed the `model` module and it looks good.  I have added a couple of minor comments but nothing that would stop this getting merged.
I liked you solution to reading the avro partition information.  I'm going to make time to review the other modules over the next week.

File: src/model/manifest.rs [hunk 2]
```diff
@@ -0,0 +1,1509 @@
...
+    }
+    /// Partition data tuple, schema based on the partition spec output using partition field ids for the struct field ids
+    pub fn partition_values(&self) -> &PartitionValues {
+        match self {
+            ManifestEntry::V1(entry) => &entry.data_file.partition,
+            ManifestEntry::V2(entry) => &entry.data_file.partition,
+        }
+    }
+    /// Full URI for the file with a FS scheme.
+    pub fn file_path(&self) -> &str {
```
[the remark this answers is not in the document]
Comment (JanKaul) [reply src/model/manifest.rs:190]: Thanks for the comments. Here I'm not entirely sure what you mean? Do you mean `std::path::Path`. I couldn't find any URI crate.

File: src/model/manifest.rs [hunk 3]
```diff
@@ -0,0 +1,1509 @@
...
+                    let result = acc? + &x?;
+                    Ok(result)
+                },
+            )?
+            .trim_end_matches(',')
+            .to_owned()
+            + r#"]}"#)
+    }
+    /// Iterate over values
+    pub fn iter(&self) -> std::slice::Iter<'_, Option<Value>> {
```
[the remark this answers is not in the document]
Comment (JanKaul) [reply src/model/manifest.rs:332]: If I understood correctly `IntoIterator` takes ownership of the values. I wanted something similar to `iter()` on Vec. I updated the implementation to use `Deref` for `PartitionValues` as `[Option<Value>]`. This way one can simply use the `iter()` method of `[T]` and it's more idiomatic the the std traits.

**Comment by timvw** [comment]
@JanKaul Do you also have an iceberg table with relative paths? Or is this not possible in  an iceberg table?

Currently trying load a datafusion table [here](https://github.com/timvw/qv/blob/iceberg/src/globbing_table.rs#L30) but running into thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Object at location /home/iceberg/warehouse/nyc/taxis/metadata/snap-638933773299822130-1-7e6760f0-4f6c-4b23-b907-0a5a174e3863.avro not found: No such file or directory (os error 2)', src/globbing_table.rs:30:168

**Comment by JanKaul** [comment]
Hi @timvw, I ran into the same issue. But as you pointed out, iceberg currently only supports full paths. As a result, just copying a table is quite troublesome. Especially because the paths in the manifest_list and manifest files need to be updated, these are avro files that cannot be simply changed by hand. I guess the best solution would be to have a DataFusion Writer which can write the iceberg table. However, I couldn't figure out a clear way to implement this. There isn't a trait for writing tables like the `TableProvider` trait for reading tables. If you have any insights, please let me know.

**Comment by timvw** [comment]
I have created a couple of my own iceberg tables with spark and noticed that it only created v1.metadata.json (and v2, v3, ..).
The current code seems to expect that there is always a v0.metadata.json which results in an exception...

**Comment by JanKaul** [comment]
Thanks for bringing up the issue. If I understand you correctly the problem is that iceberg-rs creates tables that start with v0.metadata.json but in order to be compatible with Spark it should start with v1.metadata.json. I created a fix in the branch: https://github.com/JanKaul/iceberg-rs/tree/fix/sequence_number. It would be cool if you could have a look at it

**Comment by timvw** [comment]
> fix/sequence_number

Hello Jan, 
Apologies for the distraction (I passed an incorrect path) so it also works with zero as initial sequence number.
I noticed that there is also a "version-hint.text" file which contains the "highest" vXXX.metadata.json

**Comment by timvw** [comment]
When debugging a bit further, I notice that in get_manifests the manifest_list function returns something such as "s3://bucketname/xxx" which is then passed to object_store.get (and fails, because the s3://bucketname needs to be stripped).

When i do strip that prefix, I get an iceberg table:
metadata with location: s3://bucket/path/xxx
metadata_location: path
manifests: entries with manifest_path: s3://bucket/path/xxx

When trying to use the table, I get the following error: Error: External(InvalidPort)
        let object_store_url = ObjectStoreUrl::parse(
            "iceberg://".to_owned() + &self.metadata().location().replace('/', "-"),
        )?;

**Comment by JanKaul** [comment]
Again, thanks for bringing up the issue. To be honest I haven't tried to run a query against s3. I will try to make a test case and have a look. For the "object_store_url" I wasn't entirely sure what to do.

I'm not sure if this is relevant but if I'm not mistaken it is not possible to have filesystem tables on s3 because s3 doesn't provide an atomic "copy if not exist" operation. This should be the same for all query engines.

**Comment by timvw** [comment]
Here are the steps I took to achieve this:
* I use minio to host my (local) s3 bucket: https://github.com/timvw/arrow-testing
* (I created the iceberg table, using s3a with the code here: https://github.com/timvw/demo-spark-iceberg)

In datafusion I register an object_store for s3a and try to query it. 

Hope this helps a bit ;)

**Pull request closed without merging**